### DISCUSSION TITLE
Stuck on image scaling

### DISCUSSION THREAD

**Discussion by chipsenkbeil** [3rd/image.nvim#231] [Q&A]
I'm working on finishing up a talk for neovimconf this year, and one thing I'm doing is capturing a screenshot of a webpage to show in neovim, which I display using this plugin.

I can't seem to get the screenshot to properly fill the window I'm creating. If I supply a geometry, I can make the image smaller, but what you see here is as big as I can get it. Is there support for filling a buffer with a single image? Any tips on how to do this? I can confirm that the image is bigger at (Pixel Height: 683, Pixel Width: 1,366) than what is shown. So it isn't the case that the screenshot is actual size.

![display-webpage-in-neovim](https://github.com/user-attachments/assets/5a8ef7c3-e984-46cc-a88a-66a15499d5af)

### Code Example

I've tried creating the image and rendering with custom geometries. I've tried hijacking the buffer. Doing an image setup with maximum height set to `nil` doesn't do anything. Same if I set the height percentage to `100`.

```lua
local api = require("image")
local image_path = "path/to/image.png"

-- Create a scratch buffer that is wiped once hidden
local buf = vim.api.nvim_create_buf(false, true)
vim.api.nvim_buf_set_option(buf, "bufhidden", "wipe")
vim.api.nvim_buf_set_keymap(buf, "n", "<ESC>", "<CMD>close<CR>", { noremap = true, silent = true })
vim.api.nvim_buf_set_keymap(buf, "n", "q", "<CMD>close<CR>", { noremap = true, silent = true })

-- Create a floating window using the scratch buffer positioned in the middle
local height = math.ceil(vim.o.lines * 0.8)  -- 80% of screen height
local width = math.ceil(vim.o.columns * 0.8) -- 80% of screen width
local win = vim.api.nvim_open_win(buf, true, {
    style = "minimal",
    relative = "editor",
    width = width,
    height = height,
    row = math.ceil((vim.o.lines - height) / 2),
    col = math.ceil((vim.o.columns - width) / 2),
    border = "single",
})

-- Load the image from disk and display it within our floating window
local image = api.from_file(image_path, { window = win, buffer = buf })

-- Render the image into the buffer
image:render()

-- Change to the window that is floating
vim.api.nvim_set_current_win(win)
```

**Comment by 3rd** [comment]
Hey, sorry for the trouble, could you try setting `max_height_window_percentage = 100` and checking if that works?
Doing this should be enough:
```lua
require("image").hijack_buffer("file.png", winnr, bufnr)
```

**Reply by chipsenkbeil** [reply]
Hey there! Tried that during normal setup and nothing happened. I had to override on the image prior to rendering. And set the width and height to 10000 for it to scale. 

Another question, is there a way to render part of an image? For pages that are larger than the buffer, I want to fit the width and provide keybindings to act like page up and down to show parts of the screenshot.

```lua
local function hack_image_render(image)
    local old_max_width = image.global_state.options.max_width
    local old_max_height = image.global_state.options.max_height
    local old_max_width_window_percentage = image.global_state.options.max_width_window_percentage
    local old_max_height_window_percentage = image.global_state.options.max_height_window_percentage

    -- Clear out maximum limitations so we can fill the buffer
    image.global_state.options.max_width = nil
    image.global_state.options.max_height = nil
    image.global_state.options.max_width_window_percentage = nil
    image.global_state.options.max_height_window_percentage = nil

    local old_image_width = image.image_width
    local old_image_height = image.image_height

    -- Force image dimensions to be bigger to fill the buffer
    image.image_width = old_image_width * 1000
    image.image_height = old_image_height * 1000

    image:render()

    -- Reset everything back to normal
    image.image_width = old_image_width
    image.image_height = old_image_height
    image.global_state.options.max_width = old_max_width
    image.global_state.options.max_height = old_max_height
    image.global_state.options.max_width_window_percentage = old_max_width_window_percentage
    image.global_state.options.max_height_window_percentage = old_max_height_window_percentage
end
```

**Reply by 3rd** [reply]
Hmm weird that it didn't work, it should fill the window :|
![image](https://github.com/user-attachments/assets/d96bf791-1735-4e7b-b443-f182661105de)

You should be able to use negative x/y, and it's probably better not to attach the image to a window/buffer, to get rid of some of the logic that adjusts the width/height, but even then it will cap the width/height to the maximum screen values so you'll need a hack like what you've already done.

Very interesting use case, if you use Kitty it should handle the cropping by itself and it should be pretty snappy.

![image](https://github.com/user-attachments/assets/388572e8-af90-474c-a063-e6540ea6823e)

**Reply by 3rd** [reply]
Do you want it to scroll in those directions or jump to the next/prev slide?

**Reply by chipsenkbeil** [reply]
> Do you want it to scroll in those directions or jump to the next/prev slide?

The screenshot represents the entire page, so you can imagine they can sometimes be lengthy.

If I could load the entire image into the window, fit to width, and let folks scroll that would be nice. At minimum, I'd render part of the image, and a call to page up/down would advance the height of the window in either direction.

I'm using wezterm, but had the exact same behavior on kitty. I can share the repo with you later today - I have a minimal example you load via `nvim -u scripts/003_webview.lua` - if taking a look at the full source would help.

**Reply by 3rd** [reply]
Yep, please share <3 
Are you rendering a web view (something like https://github.com/chase/awrit)?

**Reply by chipsenkbeil** [reply]
> Yep, please share <3 

Added you as a collaborator to https://github.com/chipsenkbeil/neovimconf-2024-talk

You can run the example by doing `nvim -u scripts/003_webview.lua` from the root of the repository and then run `:Webview https://chipsenkbeil.com` or some other website. Make sure it's the full address including `https`. This is demo code, so it breaks fairly easily. 😄 

Oh, and you need to clone with submodules because I have image.nvim as a submodule since it's a dependency for the 3rd example.

This is the particular code I've been referencing: https://github.com/chipsenkbeil/neovimconf-2024-talk/blob/main/scripts/003_webview/webview/lua/webview/viewer.lua

> Are you rendering a web view (something like https://github.com/chase/awrit)?

No, I'm running headless firefox using marionette. I connect over TCP via neovim's luv TCP interface and communicate via the marionette protocol to take a screenshot. I get the screenshot as base64 encoded, so I save it to a file to load up using image.nvim.

**Reply by 3rd** [reply]
Hmm, first thanks for this, got around to fixing partial rendering while scrolling.
To make scroll work we use extmarks to inject virtual space, but when scrolling Neovim will jump directly to the end of the extmark instead of smooth scrolling.
Maybe there is a way to make it not do that? It's either a missing functionality or a bad default.

https://gist.ro/browser.mp4

**Reply by 3rd** [reply]
We can connect on Discord if you want.

**Reply by chipsenkbeil** [reply]
> https://gist.ro/browser.mp4

Very nice! Would love to see what you did.

> We can connect on Discord if you want.

Sure, I can do that.
